Spark: Fix QueryFailure when running RewriteManifestProcedure on Date partitioned table

singhpk234 opened the pull request:
### About the change

Solves https://github.com/apache/iceberg/issues/5104

Presently when `spark.sql.datetime.java8API.enabled` is set to true, 

- java.time.LocalDate is returned for Spark SQL DATE type
- java.time.Instant is returned for Spark SQL TIMESTAMP type

so always casting date type to java.sql.Date or timestamp to java.sql.Timestamp is not correct and can lead to class cast exception. This change attempts to use appropriate api's from DateTimeUtils class which take these scenarios into consideration and calls the relevant api's as per the object type.

```scala
  /**
   * Converts an Java object to days.
   *
   * @param obj Either an object of `java.sql.Date` or `java.time.LocalDate`.
   * @return The number of days since 1970-01-01.
   */
  def anyToDays(obj: Any): Int = obj match {
    case d: Date => fromJavaDate(d)
    case ld: LocalDate => localDateToDays(ld)
  }


  /**
   * Converts an Java object to microseconds.
   *
   * @param obj Either an object of `java.sql.Timestamp` or `java.time.Instant`.
   * @return The number of micros since the epoch.
   */
  def anyToMicros(obj: Any): Long = obj match {
    case t: Timestamp => fromJavaTimestamp(t)
    case i: Instant => instantToMicros(i)
  }
```

----

### Testing Done 

Added an UT which would fail without this change.

cc @rdblue @aokolnychyi

ajantha-bhat commented on spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteManifestsProcedure.java:
```diff
@@ -75,6 +90,57 @@ public void testRewriteLargeManifests() {
   }
 
+  @Test
+  public void testRewriteLargeManifestsWithDatePartitionedTables() {
```
Nice PR. 

Can we please also add a timestamp testcase as the code modification touches that area? (maybe the same test case with parameterised inputs)?

singhpk234 replied:
added an UT for timestamp as well.

ajantha-bhat commented on spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteManifestsProcedure.java:
```diff
@@ -75,6 +96,79 @@ public void testRewriteLargeManifests() {
+  public void testRewriteLargeManifestsWithDateOrTimestampPartitionedTables() {
+    String[] scenarios = new String[] {"true", "false"};
+    String[] partitionColTypes = new String[] {"DATE", "TIMESTAMP"};
+    for (String scenario : scenarios) {
```
Thanks for adding now. Testcase cover all the scenarios. 

Instead of these two for loops, we can use Junit's `ParameterizedTest` with `CsvSource`, it will be very clean compared to this. (as we already depend on Junit5)

Example:
https://github.com/projectnessie/nessie/blob/main/clients/spark-extensions-basetests/src/main/java/org/projectnessie/spark/extensions/AbstractNessieSparkSqlExtensionTest.java#L589-L596

singhpk234 replied:
> Instead of these two for loops, we can use Junit's ParameterizedTest with CsvSource, it will be very clean compared to this. (as we already depend on Junit5)

+1 on the suggestion

I tried making this UT a parameterized test as per last feedback, but this UT didn't ran when i ran the complete suite, I think this is happening due to `SparkCatalogTestBase` already running with a parameterized class and we extending that.
https://github.com/apache/iceberg/blob/a8b67085e671bb5fa5a7f738ce2e9cb2dcc65d9c/spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/SparkCatalogTestBase.java#L27

Let me dig in more.

rdblue replied:
I'd be fine using `@ParamterizedTest`, but it would also work to just have two test cases that both enable Java 8 date/time types.

ajantha-bhat reviewed (approved):
Just a minor nit comment about testcase refactoring. PR is LGTM. Great finding and fix.

aokolnychyi commented:
I'll take a look over the weekend.

rdblue commented on spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteManifestsProcedure.java:
```diff
@@ -75,6 +96,79 @@ public void testRewriteLargeManifests() {
+                                "facilities",
+                                partitionColumn(partitionColType, "2021-01-04"))),
+                        new StructType(
+                            ("DATE".equals(partitionColType) ? dateStruct : timeStampStruct)))
```
Can this load the table type from the table rather than creating a static one? Something like this:

```java
sparkSchema = spark.table(tableName).schema()
```

rdblue commented on spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRewriteManifestsProcedure.java:
```diff
@@ -75,6 +96,79 @@ public void testRewriteLargeManifests() {
+  @Test
+  public void testRewriteLargeManifestsWithDateOrTimestampPartitionedTables() {
+    String[] scenarios = new String[] {"true", "false"};
+    String[] partitionColTypes = new String[] {"DATE", "TIMESTAMP"};
```
I don't think we really need to test the case where Java 8 date/time classes are disabled. That's the normal case that we know works, right? I guess it's good to have for completeness?

singhpk234 replied:
> That's the normal case that we know works, right? I guess it's good to have for completeness?

+1, the normal case works, added the normal case for increasing the coverage and completeness as we didn't had UT's for date / timestamp partitioned tables before.

singhpk234 replied:
removed the ut's for java8 date/ time disabled, and seperated out timestamp & date partitioned table UT into diff methods.

rdblue reviewed (changes requested):
The changes look okay, but the tests need to be more direct. I don't like running 4 tests in a loop inside one method when 2 test methods would do fine. Can you please remove the complexity of parameterizing this test?

singhpk234 commented:
>  I don't like running 4 tests in a loop inside one method when 2 test methods would do fine. Can you please remove the complexity of parameterizing this test?

ACK, removed the loops and added 2 test methods for date partitioned table with datetime java8 enabled and timestamp partitioned table with datetime java8 enabled.

aokolnychyi commented:
Thanks, @singhpk234! Could you, please, cherry-pick these changes to 3.2 and 3.1?

singhpk234 commented:
Thanks @aokolnychyi @rdblue @ajantha-bhat !!!

Added a pr for backporting it to 3.2 and 3.1 https://github.com/apache/iceberg/pull/6190
